packages/router/src/active-route-loader.tsx

```diff
@@ -0,0 +1,260 @@
+import React, { useRef, useState, useContext, useEffect } from 'react'
+
+import { unstable_batchedUpdates } from 'react-dom'
+
+import { Spec } from './util'
+import {
+  createNamedContext,
+  getAnnouncement,
+  getFocus,
+  resetFocus,
+} from './util'
+
+import { ParamsProvider, useLocation } from '.'
+
+const DEFAULT_PAGE_LOADING_DELAY = 1000 // milliseconds
+
+export interface PageLoadingContextInterface {
+  loading: boolean
+}
+
+export const PageLoadingContext =
+  createNamedContext<PageLoadingContextInterface>('PageLoading')
+
+export const usePageLoadingContext = () => {
+  const pageLoadingContext = useContext(PageLoadingContext)
+
+  if (!pageLoadingContext) {
+    throw new Error(
+      'usePageLoadingContext must be used within a PageLoadingContext provider'
+    )
+  }
+
+  return pageLoadingContext
+}
+
+export type LoadingState = 'PRE_SHOW' | 'SHOW_LOADING' | 'DONE'
+type LoadingStateRecord = Record<
+  string,
+  {
+    state: LoadingState
+    page: React.ComponentType<unknown>
+  }
+>
+
+interface ActivePageState {
+  loadingState: LoadingStateRecord
+}
+
+const ActivePageContext = React.createContext<ActivePageState | undefined>(
+  undefined
+)
+
+export const useActivePageContext = () => {
+  const activePageContext = useContext(ActivePageContext)
+
+  if (!activePageContext) {
+    throw new Error(
+      'useActivePageContext must be used within a ActivePageContext provider'
+    )
+  }
+
+  return activePageContext
+}
+
+type synchronousLoaderSpec = () => { default: React.ComponentType<unknown> }
+
+interface Props {
+  path: string
+  spec: Spec
+  delay?: number
+  params?: Record<string, string>
+  whileLoadingPage?: () => React.ReactElement | null
+  children?: React.ReactNode
+}
+
+const ArlNullPage = () => null
+const ArlWhileLoadingNullPage = () => null
+
+export const ActiveRouteLoader = ({
+  path,
+  spec,
+  delay,
+  params,
+  whileLoadingPage,
+  children,
+}: Props) => {
+  const location = useLocation()
+  const [pageName, setPageName] = useState('')
+  const loadingTimeout = useRef<NodeJS.Timeout>()
+  const announcementRef = useRef<HTMLDivElement>(null)
+  const mounted = useRef(true)
```

Comment:
In this case it doesn't matter if it starts out being `true` or `false`.
But it's an interesting question nonetheless 🙂 

Let's say we change it to be `const mounted = useRef(false)`.
Then, further down, we set it to `true` inside a useEffect with an empty dependency array.

useEffect hooks are called **after** render. So `mounted.current` would still have been `false` on the first render. But clearly the component is mounted, otherwise it couldn't have been rendered, right? You can't render an unmounted component, right?

On the other hand, people use this to simulate `componentWillMount()` in function components like this:

```
const willMount = useRef(true)

if (willMount.current) {
  // Before mounting
}

willMount.current = false
```

So, all this to say I'm not 100% sure exactly when a function component is considered mounted.

Reply:
Let's say you have this React app

```tsx
import React, { useEffect, useRef } from 'react';
import { render } from 'react-dom';

const App = () => {
  const mounted = useRef(false)

  useEffect(() => {
    mounted.current = true

    return () => {
      mounted.current = false
    }
  }, [])

  return <p>mounted: {'' + mounted.current}</p>
}

render(<App />, document.querySelector('#app'));
```

It will render once, and it'll say "mounted: false". That feels a little weird to me 🙂